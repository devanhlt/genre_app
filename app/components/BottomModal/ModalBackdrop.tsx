import React, { useMemo } from "react"

import { BottomSheetBackdropProps } from "@gorhom/bottom-sheet"
import Animated, { Extrapolate, interpolate, useAnimatedStyle } from "react-native-reanimated"

export const ModalBackdrop = ({ style }: BottomSheetBackdropProps) => {
  // animated variables
  const containerAnimatedStyle = useAnimatedStyle(() => ({
    opacity: interpolate(1, [0, 1], [0, 1], Extrapolate.CLAMP),
  }))

  // styles
  const containerStyle = useMemo(
    () => [
      style,
      {
        backgroundColor: "rgba(0,0,0,0.3)",
      },
      containerAnimatedStyle,
    ],
    [style, containerAnimatedStyle],
  )

  return <Animated.View style={containerStyle} />
}
