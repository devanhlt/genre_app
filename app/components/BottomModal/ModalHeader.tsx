import { Typography } from "../Typography"
import { PhosphorIcon } from "app/components/Icon/PhosphorIcon"
import { appColors } from "app/theme"
import { responsiveHeight } from "app/utils/screens"
import { flatten } from "ramda"
import * as React from "react"
import { StyleProp, TouchableOpacity, View, ViewStyle } from "react-native"

const CONTAINER: ViewStyle = {
  flexDirection: "row",
  justifyContent: "space-between",
  alignItems: "center",
  borderBottomColor: appColors.palette.black100,
  borderBottomWidth: 1,
  paddingBottom: responsiveHeight(16),
  marginBottom: responsiveHeight(16),
}

export interface ModalHeaderProps {
  /**
   * An optional style override useful for padding & margin.
   */
  style?: StyleProp<ViewStyle>
  label: string
  onCLose: () => void
}

/**
 * Describe your component here
 */
export function ModalHeader(props: ModalHeaderProps) {
  const { style, label, onCLose } = props
  const styles = flatten([CONTAINER, style])

  return (
    <View style={styles}>
      <Typography text={label} preset="title02" />
      <TouchableOpacity onPress={onCLose}>
        <PhosphorIcon name="XCircle" />
      </TouchableOpacity>
    </View>
  )
}
