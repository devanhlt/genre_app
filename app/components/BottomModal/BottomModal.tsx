import { BottomSheetModal, BottomSheetModalProps } from "@gorhom/bottom-sheet"
import React, { forwardRef, useCallback, useImperativeHandle, useMemo, useRef } from "react"
import { View, ViewStyle } from "react-native"

import { BottomSheetModalMethods } from "@gorhom/bottom-sheet/lib/typescript/types"
import { ModalBackdrop } from "./ModalBackdrop"
import { ModalHeader } from "./ModalHeader"
import { spacing } from "app/theme"
import { isFunction } from "app/utils/func"

interface BottomModalProps extends BottomSheetModalProps {
  ModalFooterComponent?: JSX.Element
  style?: ViewStyle
  contentStyle?: ViewStyle
  title?: string
  id?: string
  children: React.ReactNode
  snapPoints: string[]
}

export const BottomModal = forwardRef<BottomSheetModalMethods, BottomModalProps>(
  function BottomModal(props, ref) {
    const {
      style: $containerStyleOverride,
      contentStyle: $contentStyleOverride,
      title = "",
      id = `bottom-modal-${new Date().getTime}`,
      ModalFooterComponent,
      snapPoints: bottomModalSnapPoints = ["50%"],
      children,
    } = props

    // ref
    const bottomSheetModalRef = useRef<BottomSheetModal>(null)

    // variables
    const snapPoints = useMemo(() => bottomModalSnapPoints, [bottomModalSnapPoints])

    const onCloseModal = () => {
      if (isFunction(bottomSheetModalRef?.current?.close)) {
        bottomSheetModalRef.current.close()
      }
    }

    const onDismissModal = () => null

    const renderBackdrop = useCallback((props) => <ModalBackdrop {...props} />, [])

    useImperativeHandle(
      ref,
      () => {
        return bottomSheetModalRef.current
      },
      [],
    )

    const content = children || <></>
    return (
      <BottomSheetModal
        name={id}
        backdropComponent={renderBackdrop}
        ref={bottomSheetModalRef}
        enablePanDownToClose={false}
        detached={true}
        snapPoints={snapPoints}
        onDismiss={onDismissModal}
      >
        <View style={[$modalContainer, $containerStyleOverride]}>
          <ModalHeader label={title} onCLose={onCloseModal} style={$modalHeader} />
          <View style={[$modalContent, $contentStyleOverride]}>{content}</View>
          {ModalFooterComponent}
        </View>
      </BottomSheetModal>
    )
  },
)

const $modalContainer: ViewStyle = {
  flex: 1,
}

const $modalContent: ViewStyle = {
  flex: 1,
}

const $modalHeader: ViewStyle = {
  paddingHorizontal: spacing.size16,
}
