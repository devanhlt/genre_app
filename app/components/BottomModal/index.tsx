import { BottomSheetModal, BottomSheetModalProps } from "@gorhom/bottom-sheet"
import React, { forwardRef, useCallback, useImperativeHandle, useMemo, useRef } from "react"
import { View, ViewStyle } from "react-native"

import { BottomSheetModalMethods } from "@gorhom/bottom-sheet/lib/typescript/types"
import { ModalBackdrop } from "./ModalBackdrop"
import { ModalHeader } from "./ModalHeader"
import { spacing } from "app/theme"

interface BottomModalProps extends BottomSheetModalProps {
  ModalFooterComponent?: JSX.Element
  style?: ViewStyle
  contentStyle?: ViewStyle
  title?: string
  id?: string
  children: React.ReactNode
  snapPoints: string[]
}

export const BottomModal = forwardRef<BottomSheetModalMethods, BottomModalProps>(
  function LoggingScreen(props, ref) {
    const {
      style: $containerStyleOverride,
      contentStyle: $contentStyleOverride,
      title = "",
      id = `bottom-modal-${new Date().getTime}`,
      ModalFooterComponent,
      snapPoints: bottomModalSnapPoints = ["50%"],
      children,
    } = props

    // ref
    const bottomSheetModalRef = useRef<BottomSheetModal>(null)

    // variables
    const snapPoints = useMemo(() => bottomModalSnapPoints, [bottomModalSnapPoints])

    const onCloseModal = () => {
      bottomSheetModalRef.current.close()
    }

    const onDismissModal = () => null

    const renderBackdrop = useCallback((props) => <ModalBackdrop {...props} />, [])

    useImperativeHandle(
      ref,
      () => {
        return bottomSheetModalRef.current
      },
      [],
    )

    const content = children || <></>
    return (
      <BottomSheetModal
        name={id}
        backdropComponent={renderBackdrop}
        ref={bottomSheetModalRef}
        enablePanDownToClose={false}
        detached={true}
        snapPoints={snapPoints}
        onDismiss={onDismissModal}
      >
        <View style={[$modalContainer, $containerStyleOverride]}>
          <ModalHeader label={title} onCLose={onCloseModal} style={$modalHeader} />
          <View style={[$modalContent, $contentStyleOverride]}>{content}</View>
          {ModalFooterComponent}
        </View>
      </BottomSheetModal>
    )
  },
)

const $modalContainer: ViewStyle = {
  flex: 1,
}

const $modalContent: ViewStyle = {
  flex: 1,
}

const $modalHeader: ViewStyle = {
  paddingHorizontal: spacing.size16,
}
