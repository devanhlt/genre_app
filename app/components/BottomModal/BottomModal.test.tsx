import React from "react"
import renderer, { act } from "react-test-renderer"

import { BottomModal } from "./BottomModal"
import { BottomSheetModal } from "@gorhom/bottom-sheet"

describe("<BottomModal />", () => {
  beforeEach(() => {
    jest.restoreAllMocks()
    jest.resetAllMocks()
  })

  test("renders correctly ", () => {
    const tree = renderer.create(
      <BottomModal title="Edit System Configuration" snapPoints={["70%"]}>
        <></>
      </BottomModal>,
    )
    expect(tree).toMatchSnapshot()
  })

  test("should close function to be called", () => {
    const ref = React.createRef<BottomSheetModal>()

    renderer.create(
      <BottomModal ref={ref} title="Edit System Configuration" snapPoints={["70%"]}>
        <></>
      </BottomModal>,
    )
    const close = jest.spyOn(ref.current, "close")
    act(() => {
      ref.current.close()
    })

    expect(close).toBeCalled()
  })
})
