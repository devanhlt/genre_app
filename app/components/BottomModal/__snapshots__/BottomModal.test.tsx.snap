// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BottomModal /> renders correctly  1`] = `
<View
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "alignItems": "center",
          "borderBottomColor": "#EEF0F3",
          "borderBottomWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginBottom": 26.285714285714285,
          "paddingBottom": 26.285714285714285,
        },
        {
          "paddingHorizontal": 16,
        },
      ]
    }
  >
    <Text
      preset="title02"
      style={
        [
          {},
          {
            "color": "#000F27",
          },
          {
            "fontFamily": "System",
            "fontSize": 20,
            "fontWeight": "600",
            "letterSpacing": 0.1,
            "lineHeight": 28,
          },
          undefined,
        ]
      }
    >
      Edit System Configuration
    </Text>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight="24"
        bbWidth="24"
        color="#000"
        focusable={false}
        height={24}
        meetOrSlice={0}
        minX={0}
        minY={0}
        size={24}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            [
              undefined,
              undefined,
              {},
            ],
            {
              "flex": 0,
              "height": 24,
              "width": 24,
            },
          ]
        }
        tintColor="#000"
        vbHeight={256}
        vbWidth={256}
        width={24}
      >
        <RNSVGGroup
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
        >
          <RNSVGRect
            fill={null}
            height="256"
            propList={
              [
                "fill",
              ]
            }
            width="256"
            x="0"
            y="0"
          />
          <RNSVGCircle
            cx="128"
            cy="128"
            fill={null}
            propList={
              [
                "fill",
                "stroke",
                "strokeWidth",
                "strokeMiterlimit",
              ]
            }
            r="96"
            stroke={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            strokeMiterlimit={10}
            strokeWidth="16"
          />
          <RNSVGLine
            fill={null}
            propList={
              [
                "fill",
                "stroke",
                "strokeWidth",
                "strokeLinecap",
                "strokeLinejoin",
              ]
            }
            stroke={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            strokeLinecap={1}
            strokeLinejoin={1}
            strokeWidth="16"
            x1="160"
            x2="96"
            y1="96"
            y2="160"
          />
          <RNSVGLine
            fill={null}
            propList={
              [
                "fill",
                "stroke",
                "strokeWidth",
                "strokeLinecap",
                "strokeLinejoin",
              ]
            }
            stroke={
              {
                "payload": 4278190080,
                "type": 0,
              }
            }
            strokeLinecap={1}
            strokeLinejoin={1}
            strokeWidth="16"
            x1="160"
            x2="96"
            y1="160"
            y2="96"
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  />
</View>
`;
